Move notify, notifications and introduction notifications to new paradigm

MrPetovan opened the pull request:
This is a large scale refactoring to the new paradigm. This time, the entities can actually be interacted with through the dedicated business method `setSeen` that updates its state locally before we need to persist it to the database.

During refactoring, I've been able to enable several minor improvements:
- Slightly reduced the amount of `Verb::getByID` and `Verb::getID` calls by storing the verb itself in the notification entity. That said, the whole `Verb` model should be class constants, not a database table. This is out of scope of this PR though.
- Slightly reduced the amount of database lookups by passing the whole entity object as a parameter instead of just the ID.

This refactoring is limited to the existing paradigm that @nupplaphil had created when he used notifications to showcase Domain Driven Design. There are several remaining literal SQL queries on the `intro` table that deserve to be moved to its dedicated Depository class for example.

Given the extensiveness of the changes, I wasn't able to test all the code paths, especially not the ones dealing with the Mastodon API, but I was able to successfully interact with notifications on my dev node.

I've also started to capitalize the name of variable holding Entities to separate them visually from arbitrary table row arrays.

I've also continued to group classes by general feature rather than by type from #10716. Notification-related classes are almost all under `src/Navigation/Notifications`. This helps keeping classes that work together closer in the file tree. Hopefully it helps with discovery, rather than large folders of classes with no direct relationship other than their type.

MrPetovan committed:
Add new paradigm classes for notify

- Create BaseDepository class
- Create Entity, Collection, Factory and Depository classes
- Create FormattedNotification Entity, Collection and Factory to remove business logic from Notify repository
- Create new NotificationCreationIntercepted exception to allow addons to cancel notification creation
- Remove unused frio notifications/notify.tpl template

MrPetovan committed:
Rename Model\Post\UserNotifications type constants

MrPetovan committed:
Create constants for Mastodon notification types

MrPetovan committed:
Fix method name typos in Model\Post\UserNotification

- Reformat file, improve type-hinting

MrPetovan committed:
Improve accuracy of accepted parameter types in phpDoc of Database->fetch

- Suppresses an unwarranted type error raised by static analyzers

MrPetovan committed:
Add new paradigm classes for notification and introduction notifications

- Add support for bounded select in BaseDepository

MrPetovan committed:
Move notification introductions to the new paradigm

MrPetovan committed:
Move notification to the new paradigm

MrPetovan committed:
Move notify to the new paradigm

- Remove unused frion notify template
- Update API test

MrPetovan committed:
Update main translation file after strings have moved

annando commented on src/Worker/PushSubscription.php:
```diff
@@ -58,31 +60,29 @@ public static function execute(int $sid, int $nid)
 		}
 
-		$message = Notification::getMessage($notification);
+		$message = (new Notifications\Factory\Notification(DI::logger()))->getMessageFromNotification($Notification, DI::baseUrl(), $l10n);
```
I just scrolled through the code and saw this line - what is this? Why is it so complicate?

nupplaphil replied:
@MrPetovan - add the whole class to the `src/DI.php` as a new method and add `baseUrl` as a constructor argument, so the call would be:
```php
$message = DI::notificationFactory()->getMessageFromNotification($Notification, $l10n);
```
So Dice will become responsible for creating the right dependencies :-)
At least we now would see, how a "manual" dependency injection would work :-)

nupplaphil replied:
Another point: Dice caches shared objects :-)

annando commented:
Concerning `Navigation`: Why have you chosen this name and what could be grouped there as well? And would it be an idea to create some document with further suggestions for future groups?

Concerning `verb`: This has to be a database table since there is no fixed number of verbs out there. Remote systems can always introduce new ones. And storing them as an integer id rather than as a text saves a lot if disc space.

nupplaphil commented on include/api.php:
```diff
@@ -5659,23 +5662,24 @@ function api_friendica_activity($type)
 	}
 
-	$notifications = DI::notification()->getApiList(local_user());
+	$Notifies = DI::notify()->selectAllForUser(local_user(), ['order' => ['seen' => 'ASC', 'date' => 'DESC'], 'limit' => 50]);
```
A minor concern about it: The caller of this method nevertheless needs to know the table inner structure to pass the right sorting/limitation parameters. Maybe we could "abstract" this parameters (like every Depositorys are capable of having a `limit` and the Depository itself knows how to map it to the DB. But this is just a brainstorming like idea

annando replied:
@nupplaphil how would a call look like when you wanted to provide an order (a more complex one like this one above), a "group by" and a limit (in the different variants)?

MrPetovan replied:
The theory is that you would create a Depository method for the specific use case. The Depository has full knowledge of the database structure, so you would put your complex orders and group by in the Depository method. The limit can just be a integer parameter to the method.

MrPetovan replied:
Putting money where my mouth is, I simplified `selectAllForUser` to not require database structure knowledge from the callers.

nupplaphil commented on include/enotify.php:
```diff
@@ -397,42 +396,22 @@ function notification_store_and_send($params, $sitelink, $tsitelink, $hsitelink,
-		// Notification insertion can be intercepted by an addon registering the 'enotify_store' hook
-		if (!$notification) {
+		/** @var $factory Notifications\Factory\Notify */
+		$factory = DI::getDice()->create(Notifications\Factory\Notify::class);
```
please don't call `DI::getDice()`, because it was only introduced for testing purposes, not for production purpose. Instead create a `DI::notifyFactory()` method for it.

nupplaphil commented on include/enotify.php:
```diff
@@ -583,20 +561,20 @@ function notification_from_array(array $notification)
 	}
 
-	$msg = Notification::getMessage($notification);
+	$msg = (new Notifications\Factory\Notification(DI::logger()))->getMessageFromNotification($Notification, DI::baseUrl(), $l10n);
```
as already mentioned --> use `DI::notificationFactory()` instead

nupplaphil commented on src/BaseDepository.php:
```diff
@@ -0,0 +1,163 @@
+ * Similarly, we can't make an overloaded method return type more strict until we only support PHP version 7.4 but this
+ * is less pressing.
+ */
+abstract class BaseDepository
```
I think I now got the idea about `Depository`. Funny thing is that we already had a similar discussion for DDD as well :-)
DDD sees entities as fully business objects with encapsulating the whole business logic inside itself. The `Depository` sees entities as "small" data-classes and the business logic resides outside of the class. We should just beware that "IT-Gurus" like Martin Fowler are constantly warning about this "anemic domain model" (https://martinfowler.com/bliki/AnemicDomainModel.html), but I also found some examples how an anemic domain model could make sense: https://khalilstemmler.com/wiki/anemic-domain-model/

Did I got it ?

annando replied:
Slightly off topic, but: I do hope that at the hackathon you or @MrPetovan can explain in easy words what this whole stuff means. I see you using names that I keep forgetting faster than I can read a sentence. I guess it is as if I would explain the difference between a closed cycle and an open cycle (with the differences of being oxygen rich or fuel rich) in rocket engines when all that you know about rocket engines is that on the one side fuel goes in and on the side the exhaust goes out.

nupplaphil replied:
anemic domain model (Depository): entities are some kind of type-safe readonly arrays, the Depository knows how to save/select them, the "Model" classes do the whole logic stuff

DDD (Repository): entities are "big" objects which holds the whole logic inside them (encapsualted) the Repositories know how to save/select them. So all relevant logic is bound to the entities itself which should make them more robust against changes. There is no need for model-logic classes

nupplaphil replied:
The Depository is more applicable to the current codebase as far as I understand @mrpetovan because we already have model classes for logic and arrays for data.. It's less invasive.. The DDD means that we have to re-think the whole way how and where we put logic

MrPetovan replied:
Thank you for the link. My goal isn't to have an anemic domain model. I want the Entities to perform business logic using and changing their properties. However, I want to keep static calls out of Entities until they have been replaced by the corresponding Entities/Collections/Factories/Depositories, so right now it's a little light by constraint, not by choice.

What I don't want is to be able to arbitrarily change Entity properties. Each change should be associated with a specific use case. For example, for notifications from the moment one is created, no field but "seen" needs to change from `false` to `true`, so there's exactly one Entity method which task is to do just that.

This hardens the code against unexpected behaviors because we have consistent objects from their creation that can go through a very limited set of transformations by very design.

nupplaphil commented on src/Module/Notifications/Introductions.php:
```diff
@@ -82,34 +81,34 @@ public static function content(array $parameters = [])
 		// introduction
-		/** @var Introduction $notification */
-		foreach ($notifications['notifications'] as $notification) {
+		/** @var Introduction $Introduction */
```
tbh I'm still against breaking the camelCase rule for objects. I think this will lead to heterogeneous usage because you have to know what kind of type you're creating. And I don't see any reason for it, because we use the same naming-convention for the class names itself .. I can live with it, but I don't find it intuitive :X

MrPetovan replied:
It's a convention I started using while I was refactoring arrays into objects. I would change the type of a parameter or a variable from array to Entity/Collection in one part of a file, and then I would have make sure that all the related array key accesses were converted to object property accesses. So `$notification['key']` could still exist in the rest of the file while I was doing the refactoring, but `$Notification['key']` needed to be updated to `$Notification->key`.

nupplaphil replied:
aaahh now I understand .. does that mean in case there aren't any arrays left, you would camelCase the objects as well again? Than I fully agree with it :)

MrPetovan replied:
Absolutely. During the transition I need some sort of visual cue but once we get rid of arbitrary arrays there's no reason to keep the capitalized versions.

nupplaphil commented on src/Module/Notifications/Notification.php:
```diff
@@ -78,7 +78,8 @@ public static function rawContent(array $parameters = [])
 			try {
-				$success = DI::notify()->setSeen();
+				DI::notification()->setAllSeenForUser(local_user());
+				$success = DI::notify()->setAllSeenForUser(local_user());
```
what do we do if the method above worked, but this one don't .. It would be "partly" successful ;-)
Maybe we do need a "transactional" behavior for such cases?

MrPetovan replied:
Better than "transactional", it's a "transitional" behavior, @annando said that either `notify` or `notification` is going away, can't remember which. In the mean time, there will be some parallel calls.

nupplaphil replied:
The question is are we prepared in case we do need more than one list of entities to be saved at one call and what's happening in case of failure?

MrPetovan replied:
Not in this case. The `notify` call won't be performed at all if the `notification` call fails and throw an Exception.

nupplaphil commented on src/Module/Notifications/Notification.php:
```diff
@@ -107,24 +108,31 @@ public static function content(array $parameters = [])
+					DI::notification()->setAllSeenForUser($Notify->uid, ['target-uri-id' => $Notify->uriId]);
+				}
+
+				DI::notify()->setAllSeenForRelatedNotify($Notify);
```
Same point as above: Maybe we do need a "transactional" behavior for such cases?

nupplaphil commented on src/Module/Notifications/Notification.php:
```diff
@@ -107,24 +108,31 @@ public static function content(array $parameters = [])
 		DI::baseUrl()->redirect('notifications/system');
 
-		throw new HTTPException\InternalServerErrorException('Invalid situation.');
+		return '';
```
Why did you delete the Exception? I thought this point "should never happen", so an exception would be appropriate?

MrPetovan replied:
Both of these lines, the one that was removed and the one I added, are only "needed" to make static analyzers happy since they don't know that `DI::baseUrl()->redirect` interrupts the function flow. Neither of them will ever be called because of the `exit;` in `redirect`, both lines prevent an annoying `Missing return statement` message while the Exception throwing prompts an additional `Missing @throws tag in method documentation` message. In the end, the `return` statement makes the file one line shorter and that's why I went with it. 🤷‍♂️

nupplaphil replied:
A I see, just a quick idea/note for "later": the clean way would be that we don't call redirects directly and try to route per Routing class (Requesr -> Response), so we would have one point where the process stops and not that many :)

nupplaphil commented on src/Module/Notifications/Notifications.php:
```diff
@@ -42,41 +44,35 @@ class Notifications extends BaseNotifications
 
-		// Get the network notifications
+		/** @var \Friendica\Navigation\Notifications\Factory\FormattedNotification $factory */
+		$factory = DI::getDice()->create(\Friendica\Navigation\Notifications\Factory\FormattedNotification::class);
```
as above: please create a `DI::formattedNotificationFactory()` for it

nupplaphil commented on src/Navigation/Notifications/Depository/Notification.php:
```diff
@@ -0,0 +1,141 @@
+	 * @return Entity\Notification
+	 * @throws Exception
+	 */
+	public function save(Entity\Notification $Notification): Entity\Notification
```
Shouldn't we add an abstract `abstract public function save(BaseEntity): BaseEntity` to force Depositories having a common save method? - Or did I miss a point :X

MrPetovan replied:
Unfortunately, no. Doing so prevents child `Depository` classes to declare the kind of method you commented on because PHP doesn't accept function overloading with a more precise parameter. In this case, `Entity\Notification` would be more precise than `BaseEntity` and it would trigger a parse error. This is because PHP doesn't support "generic polymorphism" if you're interested in the nitty-gritty.

This is also why `Depositories` have ugly protected `_select*` methods that child classes can only use (and not overload) in their own `select*` methods to be able to fully take advantage of type-hinting.

nupplaphil replied:
What about using interfaces instead? AFAIK they support overloading, don't they?

MrPetovan replied:
Same problem for parameter type overloading. I tried with Interfaces and it still didn't work. On the other hand it does work for return values which is why `ICanCreateFromTableRow` exists!

nupplaphil commented on src/Navigation/Notifications/Entity/Notify.php:
```diff
@@ -0,0 +1,128 @@
+
+	public function updateMsgFromPreamble($epreamble)
+	{
+		$this->msg       = Renderer::replaceMacros($epreamble, ['$itemlink' => $this->link->__toString()]);
```
We should mark this for later, because it breaks the pattern here :-) .. I don't know how many dependencies `Renderer` has, but it injects them with that call into the "small" data-classes

MrPetovan replied:
This is fair enough, the smallest way to handle this would be to have a `Renderer` object we can pass to the method, right?

nupplaphil replied:
Yes, or a more advanced way would be to ask ourself why we use render inside "small" data entities, maybe it's part of the caller to render the msg (just an idea)

MrPetovan replied:
I fully agree.

MrPetovan replied:
So I looked into this, and having the caller render the message would pretty much pare down the method to just a couple of arbitrary property changes, which I'm not too happy about it.

But it's currently part of `include/enotify` which is slated to be refactored away, just wait until I understand why it is so complicated.

nupplaphil reviewed (changes requested):
Nice .. I think I got it now :-) .. See my comment about anemic domain model :-)

MrPetovan commented:
> Concerning `Navigation`: Why have you chosen this name and what could be grouped there as well? And would it be an idea to create some document with further suggestions for future groups?

It definitely would. So far we have three main groups of features from my work: `Security` for the two-factor authentication feature, `Navigation` for the notifications because their purpose is to redirect people to content, and `Protocol` for the Diaspora contacts. There probably are others that make more sense, we can definitely move features around groups, nothing is set in stone in that regard.

MrPetovan committed:
Create new DI shorthands for notification-related factories

MrPetovan committed:
Simplify Depository\Notify->selectAllForUser parameters

- Shifts database structure knowledge from the controller to the depository where it should be

nupplaphil commented:
what about merging this PR and replacing tomorrow the whole `Repository` classes with it, so we can have a clean baseline for future PRs? There are "just" 5 Repository-Instances:
- `Introduction`
- `FSuggest`
- <strike>`Notification`</strike> (done)
- `PermissionSet`
- `Profile`

And I guess the notification part is(t ;) ) the most complex of it
